light-system-programs/programs/verifier_program_two/src/lib.rs

```diff
@@ -45,12 +58,52 @@ pub mod verifier_program_two {
             b: proof_b,
             c: proof_c,
         };
-        process_shielded_transfer(ctx, &proof, &connecting_hash)?;
-        Ok(())
+
+        let verifier_state = VerifierState10Ins::<2, TransactionConfig>::deserialize(
+            &mut &*ctx.accounts.verifier_state.to_account_info().data.take(),
+        )?;
+
+        let public_amount = Amounts {
+            sol: verifier_state.public_amount_sol,
+            spl: verifier_state.public_amount_spl,
+        };
+
+        if *ctx.accounts.verifier_state.owner == ctx.accounts.system_program.key() {
+            return err!(crate::ErrorCode::InvalidVerifier);
+        };
+
+        let mut owner_hash = hash(&ctx.accounts.verifier_state.owner.to_bytes()).to_bytes();
+        owner_hash[0] = 0;
+        let checked_inputs = [owner_hash, connecting_hash];
+        let leaves = [
+            [verifier_state.leaves[0], verifier_state.leaves[1]],
+            [verifier_state.leaves[2], verifier_state.leaves[3]],
+        ];
+
+        let nullifiers: [[u8; 32]; 4] = verifier_state.nullifiers.to_vec().try_into().unwrap();
+        let pool_type = [0u8; 32];
+        let input = TransactionInput {
+            ctx: &ctx,
+            message: None,
+            proof: &proof,
+            public_amount: &public_amount,
+            checked_public_inputs: &checked_inputs,
+            nullifiers: &nullifiers,
+            leaves: &leaves,
+            encrypted_utxos: &verifier_state.encrypted_utxos.to_vec(),
+            relayer_fee: verifier_state.relayer_fee,
+            merkle_root_index: verifier_state.merkle_root_index as usize,
+            pool_type: &pool_type,
+            verifyingkey: &VERIFYINGKEY,
+        };
+        let mut tx =
+            Transaction::<2, 2, 4, 15, TransactionConfig, LightInstruction<'info>>::new(input);
+
+        tx.transact()
     }
 }
 
-#[light_verifier_accounts(sol, spl)]
+#[light_verifier_accounts(merkle_trees, sol, spl)]
```

Comment:
why do we have a merkle_trees flag, is there a scenario where we don't want to use any Merkle trees?

Reply:
Yes, here

https://github.com/Lightprotocol/light-protocol/blob/ac03827f958ef3c8d4de995970a04878145c7010/light-system-programs/programs/verifier_program_one/src/lib.rs#L163-L178

Before this PR, we weren't applying the macro to first instructions, but now we have to - everything which instantiates and calls `Transaction` in any way has to implement `LightAccounts`. And `Transaction` is being created to instantiate the state:

https://github.com/Lightprotocol/light-protocol/blob/ac03827f958ef3c8d4de995970a04878145c7010/light-system-programs/programs/verifier_program_one/src/lib.rs#L85-L88

If we want to avoid that, we would need to instantiate the state without instantiating `Transaction`. I will check if it makes sense.